lib/class-wp-rest-block-editor-settings-controller.php

```diff
@@ -0,0 +1,267 @@
+<?php
+/**
+ * REST API: WP_REST_Block_Editor_Settings_Controller class
+ *
+ * @package    WordPress
+ * @subpackage REST_API
+ */
+
+/**
+ * Core class used to retrieve the block editor settings via the REST API.
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_Block_Editor_Settings_Controller extends WP_REST_Controller {
+	/**
+	 * Constructs the controller.
+	 */
+	public function __construct() {
+		$this->namespace = '__experimental/wp-block-editor/v1';
+		$this->rest_base = 'settings';
+	}
+
+	/**
+	 * Registers the necessary REST API routes.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Checks whether a given request has permission to read block editor settings
+	 *
+	 * @since 5.8.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|bool True if the request has permission, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) {// phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		if ( current_user_can( 'edit_posts' ) ) {
+			return true;
+		}
+
+		foreach ( get_post_types( array( 'show_in_rest' => true ), 'objects' ) as $post_type ) {
+			if ( current_user_can( $post_type->cap->edit_posts ) ) {
+				return true;
+			}
+		}
+
+		return new WP_Error(
+			'rest_cannot_read_block_editor_settings',
+			__( 'Sorry, you are not allowed to read the block editor settings.', 'gutenberg' ),
+			array( 'status' => rest_authorization_required_code() )
+		);
+	}
+
+	/**
+	 * Returns the block editor's settings
+	 *
+	 * @since 5.8.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		$context  = ! empty( $request['context'] ) ? $request['context'] : 'post-editor';
+		$settings = gutenberg_get_block_editor_settings( $context );
+
+		return rest_ensure_response( $settings );
+	}
+
+	/**
+	 * Retrieves the block editor's settings schema, conforming to JSON Schema.
+	 *
+	 * @since 5.8.0
+	 *
+	 * @return array Item schema data.
+	 */
+	public function get_item_schema() {
+		if ( $this->schema ) {
+			return $this->add_additional_fields_schema( $this->schema );
+		}
+
+		$this->schema = array(
```

Comment:
This sounds great. It looks like there are some items that are off in terms of in which context they're available. Some that I've noticed that should be `site-editor` only are: `__experimentalGlobalStylesUserEntityId`, `__experimentalGlobalStylesBaseStyles`. Other than that, it seems that we can land this PR already? The global styles settings use the old filter `block_editor_settings` and we should update it but it doesn't block this PR, as far as I'm aware.

Also: any thoughts on best practices to keep this schema up to date as things evolve would be great.

Reply:
> Also: any thoughts on best practices to keep this schema up to date as things evolve would be great.

If we would start using the same endpoint for the web then it could help to maintain the schema 😄 

In related efforts, I plan we switch to the REST API endpoint for block types on the web rather than expose them with JS function call on the server.

Reply:
Thanks @nosolosw !

> Some that I've noticed that should be site-editor only are: __experimentalGlobalStylesUserEntityId, __experimentalGlobalStylesBaseStyles

I updated those to be `site-editor` only.

> The global styles settings use the old filter block_editor_settings and we should update it but it doesn't block this PR, as far as I'm aware.

Awesome 😃! Do you know more or less when that's going to be updated?

Reply:
I've just merged https://github.com/WordPress/gutenberg/pull/31762 which adds more data for mobile. Upon hitting the button, it just occurred to me that something was off. I think it boils down to the contexts we define in this PR: `post-editor`, `site-editor`, `widgets-editor`. Shouldn't we also have a `mobile` context?

My understanding is that we use this to filter what kind of data we pass through the request. If the plan is to move the settings retrieval to use the REST request, checks like https://github.com/WordPress/gutenberg/pull/31762/files#diff-0b8e98dead5542e95159bd3c482158111d3f16637e7eef9af13373358fa42dedR91 don't make sense. However, the current contexts are also not enough. I'm thinking we actually need a `mobile` context as well. @geriux et all, do you have thoughts on this?

Reply:
> I'm thinking we actually need a mobile context as well. @geriux et all, do you have thoughts on this?

I think we might if we want specific data like the consolidated styles we added [here](https://github.com/WordPress/gutenberg/pull/31762/files#diff-0b8e98dead5542e95159bd3c482158111d3f16637e7eef9af13373358fa42dedR106) since the web editor doesn't look like it'd need that right?

Reply:
OK, prepared https://github.com/WordPress/gutenberg/pull/31819 to add a `mobile` context.